File: 0257-binary-tree-paths/0257-binary-tree-paths.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        ans = []
        def dfs(node,path):
            if not node.right and not node.left:
                path.append(str(node.val))
                ans.append(path.copy())
                path.pop()
                return
            
            path.append(str(node.val))
            if node.left:
                dfs(node.left,path)
            if node.right:
                dfs(node.right,path)
            path.pop()
            
        dfs(root,[])
        for i in range(len(ans)):
            ans[i] = "->".join(ans[i])

        return ans

            

File: 0257-binary-tree-paths/0257-binary-tree-paths.java


class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> paths = new ArrayList<>();
        traverse(root, "", paths);
        return paths;
    }

    private void traverse(TreeNode node, String path, List<String> paths) {
        if (node == null) return;
        if (node.left == null && node.right == null) { // leaf node
            paths.add(path + node.val);
            return;
        }
        traverse(node.left, path + node.val + "->", paths);
        traverse(node.right, path + node.val + "->", paths);
    }
}
